File: fizzbuzz.py

from itertools import cycle, islice
from operator import concat

div = lambda x, y: not x % y

def fb1(n):
    for i in range(1, n):
        print('fizz' * div(i, 3) + 'buzz' * div(i, 5) or i)

def fb2(n):
    for i in islice(
            (concat(*v) or i
                for i, v in enumerate(
                    zip(
                        cycle(("", "", "fizz")), 
                        cycle(("", "", "", "", "buzz"))
                    ), 1
                )
            ), n
    ): print(i)



if __name__ == '__main__':
    fb1(100)

File: fizzbuzz.hs

rules :: [(Int, String)]
rules = [(3, "fizz"), (5, "buzz")]


divisibleBy :: Integral a => a -> a -> Bool
divisibleBy n m = mod n m == 0

infix 0 <|>
(<|>) :: [a] -> [a] -> [a]
n <|> [] = n
_ <|> xs = xs

----------------------------------------------------------------

fb1 :: [String]
fb1 = 
  [
    max
    (show n) 
    (concat [fizzbuzz | (r, fizzbuzz) <- rules, n `divisibleBy` r])
    | n <- [1..]
  ]

main13 :: IO ()
main13 = mapM_ putStrLn $ take 100 fb1

----------------------------------------------------------------

fb2 :: Int -> String
fb2 n = show n <|> concat [ fizzbuzz | (r, fizzbuzz) <- rules, n `divisibleBy` r ]


main2 :: IO ()
main2 = mapM_ (putStrLn . fb2) [1..100]

----------------------------------------------------------------

fb3 :: [String]
fb3 = zipWith3 (\n f b -> n <|> (f ++ b))
              (map show [1..])
              (cycle ["", "", "fizz"]) 
              (cycle ["", "", "", "", "buzz"]) 
